### PR TITLE
BNF grammar

### PR DESCRIPTION
**Pull request by timothee-haudebourg**
Completes the BNF grammar of the TreeLDR language.

I've added a production for documentation comments that can appear before definitions using Rust-style triple slash comments.
```tldr
/// Documentation
type Foo;
```

The `IRIREF` terminal production is complete but not sound (some words recognized by the production are not IRI references). The parser for this production is as sound and complete as the parser of the [`iref`](timothee-haudebourg/iref) library (In particular, see timothee-haudebourg/iref/issues/15). So there might be some discrepancies here.

### CODE REVIEW

**Comment by sbihel** [comment]
I must be missing something, what variant of BNF grammars is is? I've checked with BNF and ABNF but they both failed to parse the grammar

**Comment by timothee-haudebourg** [comment]
I'm using the W3 [EBNF syntax](https://www.w3.org/TR/REC-xml/#sec-notation) which is often used by RDF related recommendations. I just checked ABNF, there is an RFC and it seems more widely supported (there is even syntax highlighting in Github) so I think I'll translate it into ABNF if it can help you.

**Comment by sbihel** [comment]
Oh I don't have any particular preference, I just couldn't manage to compile it. And I did try EBNF but this randomly found validator https://mdkrajnak.github.io/ebnftest/ doesn't like comments. Any way, maybe the file extension should be modified?

**Comment by timothee-haudebourg** [comment]
I just fell into the rabbit hole of all the different BNF notations, it seems that "EBNF" refers to at least three different notations. I'm using the one also used by the W3C [N-Quads syntax recommendation](https://www.w3.org/TR/n-quads/) because the terminal production are almost the same as in TreeLDR, which just so happen to be a special EBNF notation defined for the W3C XML recommendation, which might be different from the [EBNF notation defined by the W3C itself](https://www.w3.org/Notation.html), which is also different from the ISO EBNF notation...

I could use ABNF instead, as defined by [RFC 5234](https://www.ietf.org/rfc/rfc5234.txt) because it is used to define IRIs (which I may need to include in this grammar at some point). I just want a notation that will allow me to omit spaces between tokens in some way, otherwise it'll be too complicated to write and read.

**Comment by timothee-haudebourg** [comment]
Turns out translating the grammar to ABNF is pretty painful. There is no syntax for complementary char sets like in EBNF (`[^abc]`). Still think ABNF is better though, I'll be able to invoke the `IRI-reference` rule directly from RFC 3987.

File: syntax/src/grammar.abnf [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+name                       = *iunreserved
+compact-iri                = name ":" suffix
+suffix                     = *( ipchar / "/" ) [ "?" iquery ] [ "#" ifragment ]
+                           / iprivate iquery [ "#" ifragment ]
+IRI-reference-literal      = "<" IRI-reference ">"
+
+string-char                = %x0-9 / %xb / %xc / %xe-21 / %x23-5b / %x5d-effff ; any char but %x22=" %x5C=\ %xA=new line %xD=carriage return.
+string                     = %x22 *(string-char / e-char / u-char) %x22 ; %x22="
+regexp-char                = %x0-9 / %xb / %xc / %xe-2e / %x30-5b / %x5d-effff ; any char but %x2F=/ %x5C=\ %xA=new line %xD=carriage return.
+regexp                     = "/" *(REGEXP_CHAR / e-char / u-char) "/"
```
Comment (sbihel) [inline syntax/src/grammar.abnf:93]:
```suggestion
regexp                     = "/" *(regexp-char / e-char / u-char) "/"
```

File: syntax/src/grammar.abnf [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+u-char 	                   = %s"\u" HEXDIG HEXDIG HEXDIG HEXDIG
+                           / %s"\U" HEXDIG HEXDIG HEXDIG HEXDIG HEXDIG HEXDIG HEXDIG HEXDIG
+e-char 	                   = "\" ( %s"t" / %s"b" / %s"n" / %s"r" / %s"f" / %x22 / "'" / "\" )
+
+docLine                    = "///" (%x0-%x9 / %xB-%xEFFFF) *WS ; everything but %xA=new line
```
Comment (sbihel) [inline syntax/src/grammar.abnf:98]:
```suggestion
docLine                    = "///" (%x0-9 / %xB-EFFFF) *WS ; everything but %xA=new line
```

File: syntax/src/grammar.abnf [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+                           / layout-intersection
+inner-named-layout-expr    = inner-layout-expr [ LWS %s"as" LWS name ]
+inner-layout-expr          = id
+                           / reference-layout
+                           / literal
+                           / field-restriction
+                           / array-layout
+                           / "(" LWS layout-expr LWS ")"
+layout-union               = innerNamedLayoutExpr *(LWS "|" LWS innerNamedLayoutExpr)
+layout-intersection        = innerNamedLayoutExpr *(LWS "&" LWS innerNamedLayoutExpr)
```
Comment (sbihel) [inline syntax/src/grammar.abnf:69]:
```suggestion
layout-union               = inner-named-layout-expr *(LWS "|" LWS inner-named-layout-expr)
layout-intersection        = inner-named-layout-expr *(LWS "&" LWS inner-named-layout-expr)
```

File: syntax/src/grammar.abnf [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+; Type expressions.
+type-expr                  = inner-named-type-expr
+                           / type-union
+                           / type-intersection
+inner-named-type-expr      = inner-type-expr [ LWS %s"as" LWS name / LWS %s"with" LWS id ]
+inner-type-expr            = id
+                           / reference-layout
+                           / literal
+                           / property-restriction
+                           / listType
```
Comment (sbihel) [inline syntax/src/grammar.abnf:47]:
```suggestion
                           / list-type
```

File: syntax/src/grammar.abnf [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+normal-type-definition     = "{" LWS [ property-list LWS ] "}" [ %s"with" LWS inner-layout-expr ";" ]
+
+layout-description          = layout-alias / structure-definition
+layout-alias               = "=" LWS inner-layout-expr LWS ";"
+structure-definition       = [ %s"for" LWS id LWS ] "{" LWS [ field-list LWS ] "}"
+
+field-list                 = field-definition *(LWS "," LWS field-definition)
+field-definition           = doc [ attributeList LWS ] id [ LWS ":" LWS [ attributeList ] LWS layout-expr ]
+
+property-definition        = doc [ attributeLis LWS ] id [ LWS ":" LWS [ attributeList ] LWS type-expr ]
```
Comment (sbihel) [inline syntax/src/grammar.abnf:30]:
```suggestion
property-definition        = doc [ attributeList LWS ] id [ LWS ":" LWS [ attributeList ] LWS type-expr ]
```

File: syntax/src/grammar.abnf [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+; TreeLDR Document.
+treeldr-document           = *item
+
+item                       = *WS (base / use / type / layout / property) *WS
+
+base                       = %s"base" LWS IRI-reference-literal LWS ";"
+use                        = %s"use" LWS IRI-reference-literal LWS %s"as" name ";"
+type                       = doc %s"type" LWS ref LWS type-description
+layout                     = doc %s"layout" LWS ref LWS layout-description
```
Comment (sbihel) [inline syntax/src/grammar.abnf:16]: Not sure what `ref` is

File: syntax/tests/grammar.rs [hunk 7]
```diff
@@ -0,0 +1,7 @@
+#[test]
+fn parse_abnf_grammar() {
+	abnf::rulelist(concat!(
+		include_str!("../src/grammar.abnf"),
+		include_str!("../src/grammar-rfc3987.abnf")
+	)).unwrap();
```
Comment (sbihel) [inline syntax/tests/grammar.rs:6]: maybe it could also validate every `.tldr` file under `/tests`?

**Review by sbihel — APPROVED** [review]
I must admit I didn't look closely into the grammar and syntax of treeldr

File: syntax/src/grammar.abnf (hunk 5 above)
Comment (timothee-haudebourg) [reply syntax/src/grammar.abnf:30]: Should be `attribute-list` even.

File: syntax/tests/grammar.rs (hunk 7 above)
Comment (timothee-haudebourg) [reply syntax/tests/grammar.rs:6]: The `abnf` crate doesn't even parse the grammar properly. I think it doesn't implement the update [RFC 7405](https://www.rfc-editor.org/rfc/rfc7405) on case sensitive strings. And I think it's just a parser, it cannot generate a parser.

**Comment by timothee-haudebourg** [comment]
I've implemented [RFC 7405 for the `abnf` crate](duesee/abnf#21) so we can test the well-formlessness of the ABNF grammar.

File: syntax/src/grammar.abnf (hunk 6 above)
Comment (timothee-haudebourg) [reply syntax/src/grammar.abnf]: Should have been `id`. Fixed now.

**Pull request merged**